### DISCUSSION TITLE
Authenticated loaders

### DISCUSSION THREAD

**Discussion by barbinbrad** [remix-run/react-router#9327] [Q&A]
It's my understanding that all matched loaders are fetched in parallel. Is there a best practice for solving the problem that we only want to execute certain loaders if a user is authenticated?

**Comment by chadams** [comment]
would really like to kknow the answer to this as well., Looked over the documents and see no mention of this. Basically need to make sure config and auth is ran BEFORE any sub routes. right now they are run in parallel and its making fetch calls with-out having the bearer tokens

**Reply by tannerlinsley** [reply]
~~Sub loaders can wait on the promises of parent loaders (provided in the args to the loader). Or you could just hoist the promise higher up to an external store and await a single promise from every loader that needs it.~~.

> Someone linked to this issue from within the TanStack / React Location discord and I originally answered without looking at the context of the repo 🤦 😅 .

**Reply by chadams** [reply]
can you please point to the docs where this is specified? I'm not entirely following. all my loaders in the tree fire in parallel. so my GET is firing without first having the auth header information (fetched from JWT refresh endpoint) I want to make sure the Auth refresh is performed before before any other loaders below it. 

the loaders geet two args params & request, I looked over those objects and don;t see anything related to parent loaders?

**Reply by chadams** [reply]
maybe an example? I find it weird I'm the only one asking about this.

**Reply by tannerlinsley** [reply]
I am not sure if you can access parent loader promises in the new version of React Router, though if I recall correctly, this is possible in Remix, so I would assume it's somehow possible in the new React Router loader features as well. This is probably a question for the Remix/React-Router team.

**Reply by barbinbrad** [reply]
Just tested this in remix. I have a parent route that calls `redirect` if the user is not authenticated, and so from a UX perspective, it works great. But I noticed that the child loader is still firing off before the redirect and returning a 401. 

I'm not sure that it matters on the server-side, but on the client-side, I probably don't want to be exposing my API endpoints to unauthenticated users who match a URL. Definitely a design decision to make loader requests in parallel. But probably something we want to be able to control. Also curious what the team's perspective is.

**Reply by barbinbrad** [reply]
(Outside of the obvious solution of adding an authentication check to every loader)

**Reply by pgsandstrom** [reply]
Its not a problem exposing API endpoints to unauthenticated user. They could just scrape them from the js bundle either way. For me this is more a question of tidyness. Its confusing to see a bunch of failed requests in the logs, both on the frontend and on the backend.

**Reply by barbinbrad** [reply]
~~I had an idea I wanted to run past you guys. It's not at all my own. But in other frameworks (.NET, for example) -- we often use decorators to "protect" a function.~~

~~What do you think about using JavaScript decorators to do the job here (in lieu of adding the same check and redirect code to each loader)?~~

```ts
@authenticated
export const loader = () => {}
```

~~I don't have experience with JavaScript decorators yet, but I'm thinking they are just function wrappers.~~

**Reply by chadams** [reply]
what we really need is a third argument  to loader that passes in the parent loaders result, this way you can await those responses before proceeding. TanStack router does this

**Reply by supersnager** [reply]
```tsx
const router = createBrowserRouter([
  {
   path: "/users"
   element: <UsersPage />
   loader: () => {
     return fetchUsers();
   }
  }
]);

ReactDOM.createRoot(document.getElementById("root")).render(
<AuthProvider>
  <SecondProvider>
    <ThirdProvider>
      <RouterProvider  router="router" />
    </ThirdProvider>
  </SecondProvider>
</AuthProvider>);

```
I have a similar problem.
My AuthProvider doesn't render children when authState is false (or when the token is fetching), however, due to React reconciliation algorithm the router loader is called first (before the AuthProvider renders).
I'm not sure but this is probably caused by the use of the following code: https://github.com/remix-run/react-router/blob/a1ad764f14cdb8f11892995bcd405c5536e389d8/packages/react-router/lib/components.tsx#L60

After digging in the react-router docs and source I haven't found a better solution than checking auth state in the loader.
However, the problem is even worst when you have your router definition outside of the React tree because in the loader function there is no direct access to the state from the AuthProvider.

I'm trying to resolve this by wrapping the AuthProvider with the redux Provider and using the redux state directly (outside of the react tree) in the loader.

Another solution (without redux) is to save an auth state in the simple store object, set its value in the AuthProvider imperatively, and use it in the loader.
Something like this:
```ts
export const AuthStore = {
 loggedIn: false
}
```

It is possible that my way is good, but I am not sure.
I would really appreciate if someone from the react-router team could explain what is the recommended way to secure the route without downloading data through the loader.
I don't believe that this was not thought through when designing v6.4 because after all, it is one of the basic things in applications that use routing.

The new solutions used in v6.4 combined with React Relay will be awesome if this problem is solved.

**Reply by supersnager** [reply]
Here is my simple pattern: https://github.com/remix-run/react-router/discussions/9424#discussioncomment-4062525

**Comment by chadams** [comment]
Adding this here for those stumbling onto this conversation
https://github.com/remix-run/react-router/issues/9188

Might be worth nothing how other frameworks do it: 
https://react-location.tanstack.com/guides/route-loaders#parallelized-execution-vs-serial-execution

**Comment by brophdawg11** [comment]
Hey folks!  At the moment, just like Remix, all loaders are called in parallel and you should be checking stuff like auth state in all of your loaders.  This is also important because with nested routing, re-used routes will _not_ have their loaders re-executed.  So if you only check in the `/parent` loader, then going from `/parent/child-1` to `/parent/child-2` will not re-check.

We have some ideas floating around internally on ways to solve this, so we'll update this discussion as soon as we start to solidify our ideas there.

**Reply by lzm0x219** [reply]
mark🙀

**Comment by ryanflorence** [comment]
We're designing the middleware API right now that will meet this use case and will be publishing an RFC soon.

You can see its progress on the roadmap here: https://github.com/orgs/remix-run/projects/5/views/1

**Comment by danielpza** [comment]
Something like this can work:

```ts
let resolveAuthenticatedPromise: (result: unknown) => void;

let authenticatedPromise = new Promise((resolve) => {
	resolveAuthenticatedPromise = resolve;
});

/** Call this function when the user is authenticated */
export function authenticateUser() {
	resolveAuthenticatedPromise(undefined);
}

/**
 * Makes sure the user is authenticated, useful to avoid 401 calls to the api
 *
 * @example
 * 	const loader = async () => {
 * 		return defer({ data: ensureAuthenticated().then(() => fetchData()) });
 * 	};
 */
export async function ensureAuthenticated() {
	return authenticatedPromise;
}
```

**Reply by phoneticallySAARTHaK** [reply]
Why `defer` in the loader?

**Reply by danielpza** [reply]
I thought it could work without the defer but for me it just got stuck waiting for the promise to be resolved but the parent react tree never got rendered? idk

**Reply by phoneticallySAARTHaK** [reply]
I just tried something similar (using an external resolver), and it doesn't work for me.
When the promise is resolved, the loader is executed but nothing is rendered.

However, using an [Updatable Promise](https://stackoverflow.com/q/76360105/8791684), it works just fine. Either it's some javascript/promise quirk that I'm missing, or maybe it's due to way the RR handles promises in loaders. 


<details>
<summary>Snippet</summary>

```ts
let promiseResolver: (v: boolean | Promise<boolean>) => void;
const _prom = new Promise<boolean>((r) => void (promiseResolver = r));

const userChoicePromise = new UpdatablePromise(_prom, promiseResolver);

const router = createBrowserRouter(...
// root route loader:
async loader() {
  await userChoicePromise;

  //... this code gets executed once promise resolves, and element is rendered, but doesn't happen with a normal promise

  return redirect("/user-choice")
})
```
</details>

Note: The `resolver`/`userChoicePromise` is passed as prop to a react component (`fallbackElement` in my case)

@brophdawg11, Can you provide some insights?

**Reply by brophdawg11** [reply]
I'm not sure what you're running into but using a raw promise behaves as it should: https://codesandbox.io/s/admiring-dust-gjcg54?file=/src/index.js

You do not want to hinge your authentication on a single promise though - as they can only be fulfilled once and what happens if the user logs in and then logs back out?  Or their session times out and they need to be downgraded to an anonymous user?  Once your `authenticatedPromise` is resolved it's resolved _forever_.  Instead, you should be checking some source of truth each time in each of your authenticated loaders.  And eventually, middleware will provide a much simpler API for this to reduce the duplication across loaders.

**Reply by phoneticallySAARTHaK** [reply]
My bad, I should have tested on a minimal example before posting the issue.
I did test it worked as expected.
Most probably there was some unupdated loader (awaiting a promise that would never resolve) which was stalling the whole render.

Using a helper function is more convenient 

```ts
export const createAuthenticatedLoader = (loader) => {
  if (await ensureSession()) return loader();
  else throw redirect('/login');
};
```

**Reply by knownasilya** [reply]
Took the idea above and fleshed it out

```ts
export const authenticatedLoader = <T extends LoaderFunction<any>>(
  load: T
): LoaderFunction<T> => {
  return async (args: LoaderFunctionArgs) => {
    const authenticated = await authenticate();

    if (authenticated) {
      return await load(args);
    } else {
      throw redirect('/login');
    }
  };
};
```

Used like so:

```ts
export const loader = authenticatedLoader(async ({params}) => {
  const users = await api.getUsers(params.companyId);

  return {users};
});
```

Thinking of even passing a second argument with the current user to the provided function..

```ts
export const loader = authenticatedLoader(async ({params}, {currentUser}) => {
  const posts = await api.getUserPosts(currentUser.id);

  return {posts};
});
```

**Reply by vaynevayne** [reply]
hi @knownasilya   What you mean is that each loader judges its isLogin separately, and if you want to implement the loader in children and wait for the parent component loader to complete, what should you do? For example, I want to determine whether I have page access rights in the loader,
Also, where did the currentUser in the second example come from? not understand

```jsx
const routes = [
{
path:'/',
loader:authenticatedLoader(async ()=> getUsers()),
children:[
{
path:'/posts'
loader:authenticatedLoader(async ()=> getUsersPosts()) //  i want wait getUsers loader resolve, how to do it ? 
}
]
}
]
```

**Reply by knownasilya** [reply]
Sorry, I think I had something like this:

```ts
export const authenticatedLoader = <T extends LoaderFunction<any>>(
  load: T
): LoaderFunction<T> => {
  return async (args: LoaderFunctionArgs) => {
    const authenticated = await authenticate();

    if (authenticated) {
      const currentUser = await getCurrentUser();

      return await load(args, { currentUser });
    } else {
      throw redirect('/login');
    }
  };
};
```

On the other question, I have not found a way to get nested loader routes  to depend on each other. Your best bet for now would be to use some sort of cache store like described in https://tkdodo.eu/blog/react-query-meets-react-router

**Reply by vaynevayne** [reply]
thanks

**Comment by vaynevayne** [comment]
How does everyone do the permissions module?
Here's what I'm thinking. I'd like you to comment on it.

```jsx
export const routers= [
  { path: '/', element: <Navigate to="/realtime" replace /> },
  { path: '/login', element: <Login /> },
{
  element: <BasicLayout />,
children:[]
},
  { path: '*', element: <NotFound /> },
]

const BasicLayout =()=> {
const {userInfo, isLoading,fetchUserInfo} =  useStore() // zustand 
let location = useLocation();
 useEffect(() => {
    if (!localStorage.getItem('token')) {
      navigate('/login')
    }
   if(!userInfo){
     fetchUserInfo()
   }
   //... also fetch other global data

  }, [location.key])

if (isLoading  || otherGlobalDataLoading) {
    return <LoadingOverlay  />
  }

return  <MyErrorBoundary>
          <Suspense fallback={<Loader />}>
            <Outlet  />
          </Suspense>
        </MyErrorBoundary>

}

const App = ()=> {
return  <Suspense>
                <RouterProvider router={_router} />
              </Suspense>
}



// axios.ts 
import { router } from './App'
 router.navigate('/login') // token 401
localstorage.delete('token') //  option

// Login.tsx
//  Refresh the page directly, this is the easiest way, otherwise you have to reset all the state, e.g. userInfo globalData
  const doLogin =>     window.location.replace('/') 
```

**Reply by brophdawg11** [reply]
I would recommend against doing authentication checks/redirects in the components when using a `RouterProvider`, and instead do your checks in `loader`'s and login via `action`'s: https://github.com/remix-run/react-router/tree/main/examples/auth-router-provider

>  Refresh the page directly, this is the easiest way, otherwise you have to reset all the state, e.g. userInfo globalData

This is a non-issue if you use loaders since React Router will handle it all for you.  Return any user data in loaders when logged in, and on an `action` logout React Router will automatically re-run those loaders which will remove any of the stale logged-in data.

**Reply by vaynevayne** [reply]
@brophdawg11  You are right, but I still have a question, this example is missing the code to get the user information when entering the page, I consider to get the user information in the rootLoader, but the loaders are in parallel, when the rootLoader is getting the data, and the other loader executes the redirect, won't it be jumping to the login page every time?
https://stackblitz.com/edit/github-se4d2d?file=src%2FApp.tsx

**Reply by brophdawg11** [reply]
It depends on whether the user is logged in and whether the page requires authentication.

* Not logged in, accessing `/` - no redirect, root's `useLoaderData().user` will be `null`
* Not logged in, accessing `/protected` - redirects to `/login`
* Logged in, accessing `/` - no redirect, root's `useLoaderData().user` will be populated
* Logged in, accessing `/protected` - no redirect, root's `useLoaderData().user` will be populated

Because all loaders run in parallel, you should be checking authorization in every protected loader and redirecting if not logged in.  Most of the time this means just a one line utility function:

```js
function requireLoggedIn() {
  if (!isLoggedIn()) {
    throw redirect('/login');
  }
}
```

Then you can do this in every protected loader:

```js
function loader() {
  requireLoggedIn();
  // If it didn't throw the redirect we know at this point we're logge din
}
```

**Reply by vaynevayne** [reply]
https://codesandbox.io/p/github/vaynevayne/react-auth
I have made a project that can be used in production environment according to your idea, please help me to check whether my design is reasonable, in addition, clicking logout will cause an error, can you help me to solve it? thank you

**Reply by mhmdjaw** [reply]
@brophdawg11 What if the  `isLoggedIn()` contains an API call to check for authentication status? That would be too expensive to run in every loader. What I would want is to make that call once initially when the app opens and before all loaders run, and save that user object, and then in every loader I can check for authorization by checking the saved object. If the user was not logged in, and logs in at a later point, I can just reassign the new user through an action (same thing for log out).

My point is, the initial call before all loaders run is a separate matter from the middlewares that check the saved user object. I don't want all loaders to make API calls in parallel initially, hence the importance of something like `beforeLoader`.

**Reply by brophdawg11** [reply]
> What I would want is to make that call once initially when the app opens and before all loaders run, and save that user object, and then in every loader I can check for authorization by checking the saved object

There's nothing stopping you from executing code before the initial loaders run - just handle your user initialization before calling `createBrowserRouter`:

```js
import { initUser } from './user';

initUser().then(()=> {
  let router = createBrowserRouter(routes);
  let root = ReactDOM.createRoot(document.getElementById("root"))
  root.render(<RouterProvider router={router} />);
});
```

Then you can hold the user in a module scoped var in `user.js` and export utils like `getUser`/`destroyUser`/etc. to manage it from actions.

**Reply by AsuraKev** [reply]
its 2024 and i just got this situation as well. The thing is I am using Auth0 for authentication and i cant even verify that if a user is logged in or is authenticating when the loader runs .....  I felt like the loader only needs to run before mounting that page component rather than run as soon as <RouterProvider/> is mounted

**Comment by DavidRnR** [comment]
I've found a good approach I think. **Don't import your routes until your App is ready. Import your AppRoutes component lazy/Suspense**

Edited  10/24/2024: Check the version 2 without lazy/Suspense  [#version2.](https://github.com/remix-run/react-router/discussions/9327#discussioncomment-11043340)

```typescript

import { lazy, Suspense } from 'react';
import AppTheme from '@components/layout/theme/AppTheme';
import AppSnackbar from './components/ui/AppSnackbar';
import './styles/App.css';
import LoadingApp from '@components/ui/LoadingApp';
import { useInitApp } from '@hooks/useInitApp';

// Don't load AppRoutes until Auth/App is loaded
const AppRoutes = lazy(() => import('./routes/AppRoutes'));

const App = () => {
  const { isLoaded } = useInitApp();

  if (!isLoaded) {
    return <LoadingApp />;
  }

  return (
    <AppTheme>
      {/* Use Suspense for lazy AppRoutes - So loaders are not going to be called */}
      {/*  const router =   createBrowserRouter([...]) inside AppRoutes will be defined when the app is ready. */}
      <Suspense>
        <AppRoutes />
      </Suspense>

      <AppSnackbar />
    </AppTheme>
  );
};

export default App;
```

Then your AppRoutes component should be like this. I'm using **lazy Component** by react router as well. You can use `element: <div>Component</div>` if you want to.

```typescript
import { RouterProvider, createBrowserRouter } from 'react-router-dom';
import AppLayout from '@components/layout/AppLayout/AppLayout';
import { isLoggedGuard, loaderAuthGuard } from './guards';
import { tcLoader } from '@loaders/tcLoader';
import { loadDashboard } from '@loaders/dashboardLoader';

const router = createBrowserRouter([
  // PRIVATE ROUTES
  {
    element: <AppLayout />,
    loader: loaderAuthGuard(),
    children: [
      {
        path: '/dashboard',
        lazy: () =>
          import('@pages/dashboard/DashboardPage').then((module) => ({
            Component: module.default,
          })),
        loader: loaderAuthGuard(loadDashboard),
      },
      {
        path: '/user-profile',
        lazy: () =>
          import('@pages/userprofile/UserProfilePage').then((module) => ({
            Component: module.default,
          })),
        loader: loaderAuthGuard(),
      },
    ],
  },
  {
    path: '/tc',
    lazy: () =>
      import('@pages/tc/TermsConditionsPage').then((module) => ({
        Component: module.default,
      })),
    loader: loaderAuthGuard(tcLoader),
  },

  // PUBLIC ROUTES
  {
    path: '/',
    lazy: () =>
      import('@pages/welcome/WelcomePage').then((module) => ({
        Component: module.default,
      })),
  },
  {
    path: '/login',
    lazy: () =>
      import('@pages/login/LoginPage').then((module) => ({
        Component: module.default,
      })),
    loader: isLoggedGuard,
  },
]);

const AppRoutes = () => <RouterProvider router={router} />;

export default AppRoutes;
```

Then the guard.ts file

```typescript
import { LoaderFunction, LoaderFunctionArgs, redirect } from 'react-router-dom';
import { WindowWithClerk } from '@models/app';

export const loaderAuthGuard =
  <T extends LoaderFunction>(load?: T): LoaderFunction<T> =>
  async (args: LoaderFunctionArgs) => {
    const isLogged = ...;

    if (isLogged) {
      return load ? await load(args) : null;
    } else {
      throw redirect('/login');
    }
  };

export function isLoggedGuard() {
  const isLogged = ...;
  if (isLogged) {
    throw redirect('/dashboard');
  }
  return null;
}
```

**Reply by brophdawg11** [reply]
It's not recommended to create your router inside a React component because it will be mounted/unmounted multiple times and this may potentially create the router multiple times, and if `router.dispose()` is used correctly to clean up the router on unmount, you may lose router state such as in flight navigations and loader data.  You want to ensure `createBrowserRouter` is called exactly one time for your app.

https://reactrouter.com/en/main/routers/create-browser-router

> Due to the decoupling of fetching and rendering in the design of the data APIs, you should create your router outside of the React tree with a statically defined set of routes. For more information on this design, please see the [Remixing React Router](https://remix.run/blog/remixing-react-router) blog post and the [When to Fetch](https://www.youtube.com/watch?v=95B8mnhzoCM) conference talk.

**Reply by brophdawg11** [reply]
If you want to dynamically add routes to your route tree, you can use [`patchRoutesOnNavigation`](https://reactrouter.com/en/6.27.0/routers/create-browser-router#optspatchroutesonnavigation)

**Reply by DavidRnR** [reply]
Nice @brophdawg11! :). I didn't know about pathchRoutesOnNavigation. I must have missed it when I read the documentation.

Anyway in my case, createBrowserRouter is only called once. When's the App is loaded after refrshing the page.  There is not re-renders. 

I will try to replicate it in a live demo.

Thank you, surely we can reach the best way to implement it.

**Reply by brophdawg11** [reply]
> There is not re-renders

I would never feel comfortable relying on this in a React application.  For example, there are re-renders (by definition) in strict mode.

**Reply by DavidRnR** [reply]
Hi @brophdawg11 !

Here's a live example: https://stackblitz.com/edit/react-router-auth-loaders?file=src%2FApp.tsx&terminal=dev

Strict Mode is rendering twice AppRoutes, but I will keep it in that way. It's not an issue in Production. I can't find any bug for now in my real projects. Apps go really good!

This is the only clean way that I've found to prevent triggering loaders before my App/Auth is ready.

Any suggestions will be welcome :)

**Reply by DavidRnR** [reply]
Here is a better version without using lazy/Suspense. Wow I've been working a lot on this trying to get the best way ;)

Here is a live example: https://stackblitz.com/edit/react-router-auth-loaders-version-2

App.tsx really simple

```typescript
import AppRoutes from './routes/AppRoutes';

const App = () => {
  return <AppRoutes />;
};

export default App;
```


```typescript
import { LoaderFunction, LoaderFunctionArgs, redirect } from 'react-router-dom';
import { useAuthStore } from '../store/useAuthStore';
import { useAppStore } from '../store/useAppStore';

async function checkAuthStatus() {
  const {
    appLoaded,
    actions: { initApp },
  } = useAppStore.getState();
  if (!appLoaded) {
    await initApp();
    const { isLogged } = useAuthStore.getState();
    return isLogged;
  }

  const { isLogged } = useAuthStore.getState();
  return isLogged;
}

export const loaderAuthGuard =
  <T extends LoaderFunction>(load?: T): LoaderFunction<T> =>
  async (args: LoaderFunctionArgs) => {
    const isLogged = await checkAuthStatus();
    if (isLogged) {
      return load ? await load(args) : null;
    } else {
      throw redirect('/login');
    }
  };

export async function isLoggedGuard() {
  const isLogged = await checkAuthStatus();
  if (isLogged) {
    throw redirect('/home');
  }
  return null;
}
```

**Reply by kennyhei** [reply]
@DavidRnR that's pretty cool solution and it fixed my problem with JWT auth, thanks 👍 I simplified it a little bit by creating just one store so here's how I did it:

authStore.ts:
```ts
import { create } from "zustand";
import { createJSONStorage, persist } from "zustand/middleware";

interface AuthStoreState {
  isLogged: boolean;
  accessToken: string | null;
  refreshToken: string | null;
}

interface AuthStoreActions {
  actions: {
    login: ({
      email,
      password,
    }: {
      email: string;
      password: string;
    }) => Promise<void>;
    logout: () => void;
    checkAuth: () => Promise<void>;
  };
}

export const authStore = create(
  persist<AuthStoreState & AuthStoreActions>(
    (set, get) => ({
      isLogged: false,
      accessToken: null,
      refreshToken: null,
      actions: {
        login: async ({ email, password }) => {
          // function that gets access & refresh token from your API
          const response = await apiGetTokens({
            email: email,
            password: password,
          });
          const { access, refresh } = response.data;
          set(() => ({
            isLogged: true,
            accessToken: access,
            refreshToken: refresh,
          }));
        },
        logout: () => {
          set(() => ({
            accessToken: null,
            refreshToken: null,
            isLogged: false,
          }));
        },
        checkAuth: async () => {
          const {
            accessToken,
            refreshToken,
            actions: { logout },
          } = get();
          if (!isExpired(accessToken)) {
            set((state) => ({
              ...state,
              isLogged: true,
            }));
            return;
          }
          if (!refreshToken) {
            logout();
            return;
          }
          try {
            // function that asks for new access token from your API
            const response = await apiRefreshToken({ refresh: refreshToken });
            const { access, refresh } = response.data;
            set(() => ({
              isLogged: !!access,
              accessToken: access,
              refreshToken: refresh,
            }));
          } catch (error) {
            void error;
            logout();
          }
        },
      },
    }),
    {
      name: "auth-store",
      storage: createJSONStorage(() => localStorage),
      partialize: (state) =>
        Object.fromEntries(
          Object.entries(state).filter(([key]) => key !== "actions")
        ) as AuthStoreState & AuthStoreActions,
    }
  )
);

const isExpired = (token: string | null) => {
  if (!token) {
    return true;
  }
  const { exp } = JSON.parse(atob(token.split(".")[1]));
  return exp * 1000 < Date.now();
};

// this is for setting `Authorization: "Bearer " + getToken()` for API requests that require auth
export const getToken = () => {
  const { accessToken } = authStore.getState();
  return accessToken;
};
``` 

Here's the loaderAuthGuard:

```ts
export const loaderAuthGuard =
  <T extends LoaderFunction>(load?: T): LoaderFunction<T> =>
  async (args: LoaderFunctionArgs) => {
    const {
      isLogged,
      actions: { checkAuth },
    } = authStore.getState();
    await checkAuth();
    if (isLogged) {
      return load ? await load(args) : null;
    } else {
      throw redirect("/login");
    }
  };
```

I noticed that I still need to handle routes that do not have any loader function set so I have wrapped them with `ProtectedRoute.tsx` that does pretty much the same thing as `loaderAuthGuard`:

```tsx
export function ProtectedRoute() {
  const [loading, setLoading] = useState(true);
  const [auth, setAuth] = useState(false);

  useEffect(() => {
    async function isAuthenticated() {
      const {
        isLogged,
        actions: { checkAuth },
      } = authStore.getState();
      await checkAuth();
      setAuth(isLogged);
      setLoading(false);
    }

    isAuthenticated();
  }, []);

  if (loading) {
    return null;
  }

  return auth ? <Outlet /> : <Navigate to="/login" />;
}
```

**Reply by DavidRnR** [reply]
Nice @kennyhei ! 

You don't need the ProtectedRoute component for routes without loaders. Just use a loaderAuthGuard without params.

```typescript
  {
    path: '/private-path-example',
    element: <div>Private route example</div>,
    loader: loaderAuthGuard(), // No params
  }
```

Anyway ProtectedRoute component also works, but I prefer use the loaderAuthGuard function. :)

**Comment by barbinbrad** [comment]
Just regarding the original question, it seems like middleware, which is still unreleased (but supported in express), is the right approach.

**Reply by infiniteluke** [reply]
You can make your own middleware with [dataStrategy](https://reactrouter.com/en/main/routers/create-browser-router#optsdatastrategy). Put a `middleware` property on your route's [handle](https://reactrouter.com/en/main/route/route#handle) property (could be a an async function or a permissions array or whatever, and call your auth before you call the loaders/actions in `dataStrategy`. Here's an [example](https://reactrouter.com/en/main/routers/create-browser-router#middleware) the docs show

**Comment by erwinstone** [comment]
I think it could be like this:

```tsx
export function protectedLoader<T extends LoaderFunction<any>>(load: T): Awaited<T> {
  const loader = async(args: LoaderFunctionArgs) => {
    const authenticated = authStore.authenticated

    if (authenticated) {
      return await load(args)
    }
    else {
      throw redirect(config.REDIRECT.LOGIN)
    }
  }
  return loader as unknown as Awaited<T>
}
```

You still get the "type", examples of usage:

![image](https://github.com/user-attachments/assets/81edf352-0fc7-4a66-b7ad-f3b95fc47e1f)
